### PR TITLE
Promise monad

### PR DESCRIPTION
**Pull request by niwibe**

### CODE REVIEW

File: test/cats/monad/channel_spec.cljx [hunk 1]
```diff
@@ -165,3 +163,111 @@
...
+    (go
+      (let [pr1 (do-stuff 2)
+            pr2 (channel/then pr1 inc)
+            r1 (<! pr2)]
+        (t/is (= 3 r1))
+        (done)))))
+
+
+;; #+cljs
+;; (t/deftest promise-chains-with-when
```
Comment (dialelo) [inline test/cats/monad/channel_spec.cljx]: Isn't this working?

File: test/cats/monad/channel_spec.cljx [hunk 2]
```diff
@@ -165,3 +163,111 @@
...
+(t/deftest take-from-promise-twice
+  (t/async done
+    (go
+      (let [pr (do-stuff 2)
+            r1 (<! pr)
+            r2 (<! pr)]
+        (t/is (= r1 2))
+        (t/is (= r2 2))
+        (done))))
+  )
```
Comment (dialelo) [inline test/cats/monad/channel_spec.cljx]: For consistency we should take this lonely paren up, same with the ones below.

File: src/cljx/cats/monad/channel.cljx [hunk 3]
```diff
@@ -99,3 +105,56 @@
...
+       (count [_]
+         (if (nil? @state) 0 1024))
+
+       #+cljs
+       cljs.core/ICounted
+       #+cljs
+       (-count [this]
+         (if (nil? @state) 0 1))))))
+
+(defn promise
```
Comment (dialelo) [inline src/cljx/cats/monad/channel.cljx]: Should we support passing a transducer like the `chan` constructor does?

File: src/cljx/cats/monad/channel.cljx [hunk 4]
```diff
@@ -99,3 +105,56 @@
...
+       (remove! [_] @state)
+       (add!* [this item]
+         (compare-and-set! state nil item)
+         this)
+
+       #+clj
+       clojure.lang.Counted
+       #+clj
+       (count [_]
+         (if (nil? @state) 0 1024))
```
Comment (dialelo) [inline src/cljx/cats/monad/channel.cljx]: Is there any reason to be `1024` in CLJ and `1` in CLJS?

File: src/cljx/cats/monad/channel.cljx (hunk 3 above)
Comment (niwibe) [inline src/cljx/cats/monad/channel.cljx]: Hmm, you are right! I will add it.

**Comment by dialelo** [comment]
Looks good so far, ping me when you consider it ready for merge!

**Comment by dialelo** [comment]
Needs rebasing.

**Comment by niwibe** [comment]
Yes, additionally I'm working in an other approach for it...

**Pull request closed without merging**